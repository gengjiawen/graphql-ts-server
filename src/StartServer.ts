import { GraphQLServer } from 'graphql-yoga'
import { createTypeormConn } from './utils/CreateTyepeormConn'
import { genSchema } from './utils/GenSchema'

export const startServer = async () => {
  const schema = genSchema()

  const server = new GraphQLServer({ schema: schema })
  await createTypeormConn()

  const port: number = process.env.NODE_ENV === 'test' ? 9527 : 4000
  const app = await server.start(
    {
      port,
    },
    () => {
      console.log(`server is running on localhost:${port}`)
    }
  )

  return app
}
