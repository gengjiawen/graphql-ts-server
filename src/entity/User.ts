import { BaseEntity, BeforeInsert, Column, Entity, PrimaryGeneratedColumn } from 'typeorm'
import * as bcrypt from 'bcryptjs'
import { ObjectType, Field, ID } from 'type-graphql'

@ObjectType()
@Entity('users')
export class User extends BaseEntity {
  @Field(() => ID)
  @PrimaryGeneratedColumn('uuid')
  id: string

  @Field()
  @Column('varchar', { length: 255 })
  email: string

  @Field()
  @Column('text')
  password: string

  @Field()
  @Column('boolean', { default: false })
  confirmed: boolean

  @BeforeInsert()
  async hashPassword() {
    this.password = await bcrypt.hash(this.password, 10)
  }
}
