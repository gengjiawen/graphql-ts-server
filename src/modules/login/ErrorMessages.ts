export const invalidLogin = 'invalid login'
export const confirmEmailError = 'please confirm your email first'
export const accountLockedError = 'account is locked'
