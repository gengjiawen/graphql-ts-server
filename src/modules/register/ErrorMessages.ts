export const duplicateEmail = 'already taken'
export const emailNotLongEnough = 'email not long enough'
export const passwordNotLongEnough = 'password not long enough'
export const invalidEmail = 'email must be a valid email'
