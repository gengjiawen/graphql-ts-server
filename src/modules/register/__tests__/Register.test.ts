import { startServer } from '../../../StartServer'
import { request } from 'graphql-request'
import { User } from '../../../entity/User'
import { AddressInfo } from 'net'
import { duplicateEmail } from '../ErrorMessages'

let getHost = () => ''

beforeAll(async () => {
  const app = await startServer()
  const { port } = app.address() as AddressInfo
  getHost = () => `http://localhost:${port}`
})

const email = 'tom@bob.com'
const password = '123456'

const mutation = (e: string, p: string) => `
mutation {
  register(email: "${e}", password: "${p}") {
    path
    message
  }
}`
test('register user', async () => {
  const response = await request(getHost(), mutation(email, password))
  expect(response).toEqual({ register: null })
  const users = await User.find({ where: { email } })
  expect(users).toHaveLength(1)
  const user = users[0]
  expect(user.email).toEqual(email)
  expect(user.password).not.toEqual(password)

  // test duplicate email
  const response2: any = await request(getHost(), mutation(email, password))
  expect(response2.register).toHaveLength(1)
  expect(response2.register[0]).toEqual({
    path: 'email',
    message: duplicateEmail,
  })

  // catch bad password
  const response3: any = await request(getHost(), mutation(email, 'ab'))
  expect(response3).toMatchSnapshot()

  // catch bad email and password
  const response4: any = await request(getHost(), mutation('ab', 'cd'))
  expect(response4).toMatchSnapshot()
})
