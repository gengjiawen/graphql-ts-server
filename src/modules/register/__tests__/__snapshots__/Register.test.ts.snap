// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`register user 1`] = `
Object {
  "register": Array [
    Object {
      "message": "password not long enough",
      "path": "password",
    },
  ],
}
`;

exports[`register user 2`] = `
Object {
  "register": Array [
    Object {
      "message": "email not long enough",
      "path": "email",
    },
    Object {
      "message": "email must be a valid email",
      "path": "email",
    },
    Object {
      "message": "password not long enough",
      "path": "password",
    },
  ],
}
`;
